### DISCUSSION TITLE
Minimal HTTPS example with curl refuses to work

### DISCUSSION THREAD

**Discussion by slavistan** [mitmproxy/mitmproxy#6030] [General]
Hi there.

I'm trying to create a minimal working example using just mitmproxy and curl, which I am unable to get working and I'm starting to lose my mind. Why doesn't the following work?

```
mitmdump --listen-host 127.0.0.1 --listen-port 8888 --mode regular # run in another shell
curl --proxy http://127.0.0.1:8888 --proxy-insecure https://wikipedia.org
```

According to `man curl | less '+/^\s+--proxy-insecure'` I'd expect _--proxy-insecure_ to disable all SSL verification for the communication between the client (curl) and the proxy. However, something's going awry: _mitmproxy_'s responds with this:

```
[15:11:20.236][127.0.0.1:34620] client connect
[15:11:20.272][127.0.0.1:34620] server connect wikipedia.org:443 ([2620:0:862:ed1a::1]:443)
[15:11:20.305][127.0.0.1:34620] Client TLS handshake failed. The client does not trust the proxy's certificate for wikipedia.org (tlsv1 alert unknown ca)
[15:11:20.305][127.0.0.1:34620] client disconnect
[15:11:20.305][127.0.0.1:34620] server disconnect wikipedia.org:443 ([2620:0:862:ed1a::1]:443)
```

and _curl_'s verbose output is

```
*   Trying 127.0.0.1:8888...
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
* allocate connect buffer
* Establish HTTP proxy tunnel to wikipedia.org:443
> CONNECT wikipedia.org:443 HTTP/1.1
> Host: wikipedia.org:443
> User-Agent: curl/7.88.1
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 Connection established
<
* CONNECT phase completed
* CONNECT tunnel established, response 200
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
* SSL certificate problem: unable to get local issuer certificate
* Closing connection 0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

The usage of `--proxy-insecure` does not seem to make any difference at all and I don't know how to proceed to get this to work.

May I kindly ask you for help? Thank you.

**Comment by mhils** [comment]
cURL's `--proxy-insecure` disables the verification of the proxy server's own TLS certificate. This applies if you already connect to the proxy via TLS, and then possibly do TLS-over-TLS to the final destination. mitmproxy also spoofs the target server's certificate. For cURL to accept this, you either need to specify `--insecure` (shorthand `-k`), or add your mitmproxy CA to your list of trusted CAs.

In short:
 - `--insecure` for the connection to the target server.
 - `--proxy-insecure` for the connection to the proxy only (if you are using a ["Secure Web Proxy"](https://www.chromium.org/developers/design-documents/secure-web-proxy/))

or (preferably) you mark the mitmproxy CA as trusted.

**Reply by slavistan** [reply]
>mitmproxy also spoofs the target server's certificate

That's the info I needed. Thank you very much!

Here's the request which gets the job done:
```
curl --proxy http://127.0.0.1:8888 --cacert ~/.mitmproxy/mitmproxy-ca-cert.cer https://wikipedia.org
```

**Comment by frozen-cdn** [comment]
Hi there, I am having the same issue, and it's very frustrating.  Windows 11 user here.

I have fully imported the certificate using mitm.it and following the instructions to download the .p12 file, double click, import to the appropriate place and press finish.

Still, I cannot curl through the proxy anything with https:// until I explicitly declare it with --cacert and the path to the file mentioned by @slavistan above -- and then it works perfectly!

```
$ curl --proxy 192.168.1.10:8080 https://site/load.php
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
$
```
```
$ curl --proxy 192.168.1.10:8080 --cacert /users/me/.mitmproxy/mitmproxy-ca-cert.cer https://site/load.php
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL was not found on this server.</p>
<hr>
<address>Apache/2.4.29 (Ubuntu) Server at site Port 8189</address>
</body></html>
$
```
Suddenly it works.

But, why?

Isn't this the whole purpose of importing the root certificate?

Did I import the wrong one or something?

Why isn't curl obeying it?  Is there no way to ensure --cacert is not needed because the certificate is already allowed system-wide?  

So many questions..

**Reply by mhils** [reply]
> Why isn't curl obeying it?

Applications may choose to ignore the system's certificate store and use their own store instead. For cURL, it depends on the specific cURL build.
